### PR TITLE
fix:jsonrpc history

### PR DESCRIPTION
**Pull request by jorgemmsilva**
fix evm jsonrpc querying historical data on blocks that have already been pruned from the active state.

for now this relies on an in-memory index for txs and blocks by number/hash. This can be persisted to disk later (on a separate non-state-trie DB)

### CODE REVIEW

File: packages/evm/jsonrpc/jsonrpccache/cache.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+		c.blockTrieHashByIndex[i] = &blockTrieHash
+
+		evmBlock := db.GetCurrentBlock()
+		c.blockByHash[evmBlock.Hash()] = evmBlock
+
+		c.blockByNum[uint64(i)] = evmBlock
+		blockTransactions := evmBlock.Transactions()
+		c.txsByBlockHash[evmBlock.Hash()] = blockTransactions
+		for _, tx := range blockTransactions {
+			tx := tx // FIX FOR ITERATION VALUE SHARING
```
Comment (dessaya) [inline packages/evm/jsonrpc/jsonrpccache/cache.go]: Is this really needed here? `tx` isn't stored in a closure...

File: packages/vm/core/blocklog/stateaccess.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package blocklog
+
+import (
+	"github.com/iotaledger/wasp/packages/kv"
+	"github.com/iotaledger/wasp/packages/kv/subrealm"
+)
+
+type StateAccess struct {
```
Comment (dessaya) [inline packages/vm/core/blocklog/stateaccess.go:11]: `StatePartition`, `StateSubrealm`?

btw, I like this pattern of having an object to encapsulate the contract state. Let's do that with all core contracts.

File: packages/evm/jsonrpc/jsonrpccache/cache.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+func (c *Cache) CacheBlock(trieRoot trie.Hash, stateByTrieRoot func(trieRoot trie.Hash) (state.State, error)) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	state, err := stateByTrieRoot(trieRoot)
+	if err != nil {
+		panic(err)
+	}
+	blockKeepAmount := governance.NewStateAccess(state).GetBlockKeepAmount()
```
Comment (dessaya) [inline packages/evm/jsonrpc/jsonrpccache/cache.go:49]: How does this work? `NewStateAccess()` returns an instance of `*StateAccess`, and I don't see that `StateAccess` has a `GetBlockKeepAmount` method :thinking:

**Review by dessaya — APPROVED** [review]

File: packages/evm/jsonrpc/jsonrpccache/cache.go (hunk 1 above)
Comment (jorgemmsilva) [reply packages/evm/jsonrpc/jsonrpccache/cache.go]: that's right, thanks :+1:

File: packages/vm/core/blocklog/stateaccess.go (hunk 2 above)
Comment (jorgemmsilva) [reply packages/vm/core/blocklog/stateaccess.go:11]: Yeah, trend started by Karolis, I like it as well!
right now there are some implemented for core contracts and all are called `StateAccess`, so I'll keep the naming consistent for now

File: packages/evm/jsonrpc/jsonrpccache/cache.go (hunk 3 above)
Comment (jorgemmsilva) [reply packages/evm/jsonrpc/jsonrpccache/cache.go:49]: this one is from `governance` :stuck_out_tongue:

**Pull request merged**